### ISSUE TITLE
New warning in the main: Python/import.c [-Wstringop-truncation]

### ISSUE THREAD

**Issue by skirpichev** [python/cpython#127864]
# Bug report

### Bug description:

```
$ cc --version|head -1
cc (Debian 12.2.0-14) 12.2.0
$ ./configure -q && make -s
In function ‘hashtable_key_from_2_strings’,
    inlined from ‘_extensions_cache_find_unlocked’ at Python/import.c:1266:17:
Python/import.c:1179:5: warning: ‘strncpy’ output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
 1179 |     strncpy(key, str1_data, str1_len);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Python/import.c:1165:27: note: length computed here
 1165 |     Py_ssize_t str1_len = strlen(str1_data);
      |                           ^~~~~~~~~~~~~~~~~
Written build/lib.linux-x86_64-3.14/_sysconfigdata__linux_x86_64-linux-gnu.py
Written build/lib.linux-x86_64-3.14/_sysconfig_vars__linux_x86_64-linux-gnu.json
Checked 112 modules (34 built-in, 77 shared, 1 n/a on linux-x86_64, 0 disabled, 0 missing, 0 failed on import)
```

Edit: happens on non-debug builds.

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

Linux

### Linked PRs
* gh-127878
* gh-128018

**Comment by tomasr8** [comment]
Using `memcpy` instead of `strncpy` gets rid of the warning

**Comment by utkonos** [comment]
Should this issue be reopened? I observe it in the build output of 3.13.2 today:

```
    default: Python/import.c: In function ‘_extensions_cache_find_unlocked’:
    default: Python/import.c:1177:5: warning: ‘strncpy’ output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
    default:  1177 |     strncpy(key, str1_data, str1_len);
    default:       |     ^
    default: In function ‘hashtable_key_from_2_strings’,
    default:     inlined from ‘_extensions_cache_find_unlocked’ at Python/import.c:1264:17:
    default: Python/import.c:1163:27: note: length computed here
    default:  1163 |     Py_ssize_t str1_len = strlen(str1_data);
    default:       |                           ^~~~~~~~~~~~~~~~~
```

**Comment by colesbury** [comment]
It was fixed in main (3.14) but not backported to 3.13.  The warning is spurious -- there's no real bug -- so I'm not sure if it's worth backporting to 3.13.

**Comment by utkonos** [comment]
Understood. Thanks for looking! I see one other location with something similar in the build output. I'm sure you all got this one as well:

```
    default: Python/import.c: In function ‘_extensions_cache_find_unlocked’:
    default: Python/import.c:1177:5: warning: ‘strncpy’ output truncated before terminating nul copying as many bytes from a string as its length [-Wstringop-truncation]
    default:  1177 |     strncpy(key, str1_data, str1_len);
    default:       |     ^
    default: In function ‘hashtable_key_from_2_strings’,
    default:     inlined from ‘_extensions_cache_find_unlocked’ at Python/import.c:1264:17:
    default: gcc -c -fno-strict-overflow -Wsign-compare -DNDEBUG -g -O3 -Wall    -fno-semantic-interposition -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -g -std=c11 -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration -fvisibility=hidden -fprofile-use -fprofile-correction -I./Include/internal -I./Include/internal/mimalloc  -I. -I./Include    -DPy_BUILD_CORE -o Python/formatter_unicode.o Python/formatter_unicode.c
    default: Python/import.c:1163:27: note: length computed here
    default:  1163 |     Py_ssize_t str1_len = strlen(str1_data);
    default:       |                           ^~~~~~~~~~~~~~~~~
```
